ring/partition_ring_editor.go

```diff
@@ -0,0 +1,60 @@
+package ring
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/grafana/dskit/kv"
+)
+
+// PartitionRingEditor is standalone component that can be used to modify the partitions ring.
+// If you want to implement the partition lifecycle you should use PartitionInstanceLifecycler instead.
+type PartitionRingEditor struct {
+	ringKey string
+	store   kv.Client
+}
+
+func NewPartitionRingEditor(ringKey string, store kv.Client) *PartitionRingEditor {
+	return &PartitionRingEditor{
+		ringKey: ringKey,
+		store:   store,
+	}
+}
+
+// ChangePartitionState changes the partition state to toState.
+// This function returns ErrPartitionDoesNotExist if the partition doesn't exist,
+// and ErrPartitionStateChangeNotAllowed if the state change is not allowed.
+func (l *PartitionRingEditor) ChangePartitionState(ctx context.Context, partitionID int32, toState PartitionState) error {
+	return l.updateRing(ctx, func(ring *PartitionRingDesc) (bool, error) {
```

Comment:
this is copied from the `PartitionInstanceLifecycler`. Was there a specific reason for doing this? Can we instead change the `PartitionInstanceLifecycler` to delegate to an PartitionRingEditor and don't duplicate the tests and the code?

Reply:
> Can we instead change the PartitionInstanceLifecycler to delegate to an PartitionRingEditor

Unfortunately not because the lifecycler needs to run all changes through a single goroutine to avoid concurrency issues. Refactoring will probably require more code than such a little copying.

Reply:
can't the lifecycler still invoke the editor through a single goroutine?

```go 
func (l *PartitionInstanceLifecycler) ChangePartitionState(ctx context.Context, toState PartitionState) error {
	return l.runOnLifecyclerLoop(func() error {
		return l.editor.ChangePartitionState(ctx, l.cfg.PartitionID, toState)
	})
}
```

Reply:
I took a similar approach without directly using the editor. I moved the change partition logic to a dedicated function which is shared between the two: https://github.com/grafana/dskit/pull/497/commits/2e611f818a566c41cadda35d53d02f5871bdf6c9

Reply:
that looks better, but updateRing is still duplicated. updateRing is more boilerplate code, so i'm happy to keep it as-is